src/reconcile/rec_child.c

```diff
@@ -15,90 +15,80 @@
-     *
-     * First, check to see if there is an address associated with this leaf: if there isn't, we're
-     * done, the underlying page is already gone. If the page still exists, check for any
-     * transactions in the system that might want to see the page's state before it's deleted.
-     *
-     * If any such transactions exist, we cannot discard the underlying leaf page to the block
-     * manager because the transaction may eventually read it. However, this write might be part of
-     * a checkpoint, and should we recover to that checkpoint, we'll need to delete the leaf page,
-     * else we'd leak it. The solution is to write a proxy cell on the internal page ensuring the
-     * leaf page is eventually discarded.
+     * A visible entry can be in a prepared state and checkpoints skip in-progress prepared changes.
+     * We can't race here, the entry won't be visible to the checkpoint, or will be in a prepared
+     * state, one or the other.
      *
-     * If no such transactions exist, we can discard the leaf page to the block manager and no cell
-     * needs to be written at all. We do this outside of the underlying tracking routines because
-     * this action is permanent and irrevocable. (Clearing the address means we've lost track of the
-     * disk address in a permanent way. This is safe because there's no path to reading the leaf
-     * page again: if there's ever a read into this part of the name space again, the cache read
-     * function instantiates an entirely new page.)
+     * We should never see an in-progress prepare in eviction: when we check to see if an internal
+     * page can be evicted, we check for an unresolved fast-truncate, which includes a fast-truncate
+     * in a prepared state, so it's an error to see that during eviction.
      */
-    if (ref->addr != NULL && !__wt_page_del_active(session, ref, true)) {
-        WT_RET(__wt_ref_block_free(session, ref));
+    WT_ORDERED_READ(prepare_state, page_del->prepare_state);
+    if (prepare_state == WT_PREPARE_INPROGRESS || prepare_state == WT_PREPARE_LOCKED) {
+        WT_ASSERT(session, !F_ISSET(r, WT_REC_EVICT));
 
-        /* Any fast-truncate information can be freed as soon as the delete is stable. */
-        __wt_overwrite_and_free(session, ref->ft_info.del);
+        *statep = WT_CHILD_ORIGINAL;
+        return (0);
     }
 
     /*
-     * If the original page is gone, we can skip the slot on the internal page.
+     * Deal with underlying disk blocks. If there are readers that might want to see the page's
+     * state before it's deleted, or the fast-delete can be undone by RTS, we can't discard the
+     * pages. Write a cell to the internal page with information describing the fast-delete.
      */
-    if (ref->addr == NULL) {
-        *statep = WT_CHILD_IGNORE;
+    if (__wt_page_del_active(session, ref, true) ||
+      (conn->txn_global.has_stable_timestamp &&
+        page_del->timestamp > conn->txn_global.stable_timestamp)) {
```

Comment:
Also, I don't understand the need for this stable timestamp change. When the `__wt_page_del_active` returns false means that the FT durable timestamp is less than the pinned timestamp.

Reply:
I'm confused by this: I thought the visibility checks were never against the stable TS, that is, we need to check if RTS will potentially undo this committed and visible change, in which case we cannot remove the underlying blocks. Can you please explain further, or make the change you want to see?

Reply:
I think there are too many `not`s in this to reason about it safely without writing down the intermediate steps.

`__wt_page_del_active` has these cases:
1. There is no page_del structure. That is not true here so we can skip thinking about it.
2. The txnid is `WT_TXN_ABORTED`. I'm not sure that case is handled in this function at this point, also not sure if it comes up. Don't think it's what Hari's talking about though.
3. The prepare state is `INPROGRESS` or `LOCKED`. This is not true here (checked in the previous clause) so we can skip thinking about it.
4. Otherwise it returns `not` the visibility. In order for the stable timestamp check to run, `__wt_page_del_active` must return false, and since we pass `true` for `visible_all` this means that `__wt_txn_visible_all` must return true.

This in turn can happen if:
1. The connection is closing. Not sure what this implies.
2. The timestamp is `WT_TS_NONE`. In this case the additional check is always false so it doesn't add anything.
3. The timestamp is <= the pinned timestamp. In this case because pinned <= oldest and oldest <= stable, the timestamp must be less than stable so the additional check is always false and doesn't add anything.

I think Hari's point is #3. However, after checking this in detail I think we'd better keep the extra check in case the connection is closing. I don't think we want to discard the page just because we're shutting down if the deletion isn't stable yet.

Reply:
After talking to Keith about this: probably it's better to roll any closing-related logic inside `__wt_page_del_active`.

Reply:
Thanks, @sauclovian-wt for adding more details. Yes, my point is 3 as you mentioned. I got the same feeling about allowing the additional check as we don't want to discard the page during the connection closing. But later I realised that the `WT_CONN_CLOSING` flag was set only when the shutdown checkpoint is finished. 

I am not sure in which scenarios we are going to write that page again after the final checkpoint? Having that check doesn't cause any problems, just want to know any other scenario where we write this page after the final checkpoint?

Yes, I also agree to move it inside the `__wt_page_del_active` would be good.

Reply:
Thank you, @kommiharibabu, fixed in c751a47.

Reply:
@kommiharibabu In that case there's nothing to worry about. Which is a relief, because if it weren't true there are other problems the closing check might cause problems.

Reply:
Thanks.